server/etcdserver/v3_server.go

```diff
@@ -831,6 +827,59 @@ func (s *EtcdServer) linearizableReadLoop() {
 	}
 }
 
+func (s *EtcdServer) readIndexLoop() {
+	for {
+		requestId := s.reqIDGen.Next()
+		leaderChangedNotifier := s.leaderChanged.Receive()
+		select {
+		case <-leaderChangedNotifier:
+			continue
+		case <-s.readIndexWaitc:
+		case <-s.stopping:
+			return
+		}
+
+		nextnr := newNotifier()
+		s.readMu.Lock()
```

Comment:
How about adding a separate lock dedicated for read index so adding this loop has no impact on the `linearizableReadLoop`?